### PR TITLE
TS API Update - Allow reuse of particpant/publisher/subscriber with i…

### PR DESCRIPTION
**Pull request by oschwaldp-oci**
…dentical qos values.  Because multiple readers/writers (connections) can be using same participant/pub/sub need to change cleanup as well to only delete entities when no longer in use

### CODE REVIEW

File: dds/FACE/FaceTSS.cpp [hunk 1]
```diff
@@ -285,6 +306,186 @@ void Receive_Message(
...
+  void find_or_create_dp(const DDS::DomainId_t& domainId,
+                         const DDS::DomainParticipantFactory_var& dpf,
+                         DDS::DomainParticipant_var& dp) {
+    DDS::DomainParticipant_var temp_dp;
+    OPENDDS_MAP(FACE::CONNECTION_ID_TYPE, Entities::FaceReceiver)& readers = Entities::instance()->receivers_;
+    OPENDDS_MAP(FACE::CONNECTION_ID_TYPE, Entities::FaceReceiver)::iterator rdrIter = readers.begin();
+    while (rdrIter != readers.end()) {
+      temp_dp = rdrIter->second.dr->get_subscriber()->get_participant();
+      if (domainId == temp_dp->get_domain_id()) {
+        ACE_DEBUG((LM_DEBUG, "(%P|%t) find_or_create_dp - found exiting participant for domainId: %d in receivers\n", domainId));
```
Comment (iamtheschmitzer) [inline dds/FACE/FaceTSS.cpp]: protect log with 
if (DCPS_debug_level > 3)

File: dds/FACE/FaceTSS.cpp [hunk 2]
```diff
@@ -285,6 +306,186 @@ void Receive_Message(
 }
 
 namespace {
+  void find_or_create_dp(const DDS::DomainId_t& domainId,
+                         const DDS::DomainParticipantFactory_var& dpf,
+                         DDS::DomainParticipant_var& dp) {
+    DDS::DomainParticipant_var temp_dp;
+    OPENDDS_MAP(FACE::CONNECTION_ID_TYPE, Entities::FaceReceiver)& readers = Entities::instance()->receivers_;
```
Comment (iamtheschmitzer) [inline dds/FACE/FaceTSS.cpp]: Use a typedef

File: dds/FACE/FaceTSS.cpp [hunk 3]
```diff
@@ -285,6 +306,186 @@ void Receive_Message(
...
+      if (domainId == temp_dp->get_domain_id()) {
+        ACE_DEBUG((LM_DEBUG, "(%P|%t) find_or_create_dp - found exiting participant for domainId: %d in receivers\n", domainId));
+        dp = temp_dp;
+        return;
+      } else {
+        ++rdrIter;
+      }
+    }
+
+    OPENDDS_MAP(FACE::CONNECTION_ID_TYPE, Entities::FaceSender)& writers = Entities::instance()->senders_;
```
Comment (iamtheschmitzer) [inline dds/FACE/FaceTSS.cpp]: typedef

File: dds/FACE/FaceTSS.cpp [hunk 4]
```diff
@@ -285,6 +306,186 @@ void Receive_Message(
...
+      temp_dp = wtrIter->second.dw->get_publisher()->get_participant();
+      if (domainId == temp_dp->get_domain_id()) {
+        ACE_DEBUG((LM_DEBUG, "(%P|%t) find_or_create_dp - found exiting participant for domainId: %d in senders\n", domainId));
+        dp = temp_dp;
+        return;
+      } else {
+        ++wtrIter;
+      }
+    }
+    ACE_DEBUG((LM_DEBUG, "(%P|%t) find_or_create_dp - created new participant for domainId: %d\n", domainId));
```
Comment (iamtheschmitzer) [inline dds/FACE/FaceTSS.cpp]: protect

File: dds/FACE/FaceTSS.cpp [hunk 5]
```diff
@@ -285,6 +306,186 @@ void Receive_Message(
...
+    dp = dpf->create_participant(domainId, PARTICIPANT_QOS_DEFAULT, 0, 0);
+  }
+
+  void find_or_create_pub(const DDS::PublisherQos& qos,
+                          const DDS::DomainParticipant_var& dp,
+                          DDS::Publisher_var& pub) {
+    DDS::DomainParticipant_var temp_dp;
+    DDS::Publisher_var temp_pub;
+
+    OPENDDS_MAP(FACE::CONNECTION_ID_TYPE, Entities::FaceSender)& writers = Entities::instance()->senders_;
```
Comment (iamtheschmitzer) [inline dds/FACE/FaceTSS.cpp]: typedef

File: dds/FACE/FaceTSS.cpp [hunk 6]
```diff
@@ -285,6 +306,186 @@ void Receive_Message(
...
+    OPENDDS_MAP(FACE::CONNECTION_ID_TYPE, Entities::FaceSender)::iterator wtrIter = writers.begin();
+
+    while (wtrIter != writers.end()) {
+      temp_pub = wtrIter->second.dw->get_publisher();
+      temp_dp = temp_pub->get_participant();
+      DDS::PublisherQos temp_qos;
+      temp_pub->get_qos(temp_qos);
+      if (dp->get_domain_id() == temp_dp->get_domain_id() &&
+          temp_qos == qos) {
+        ACE_DEBUG((LM_DEBUG, "(%P|%t) find_or_create_pub - found exiting publisher in senders\n"));
```
Comment (iamtheschmitzer) [inline dds/FACE/FaceTSS.cpp]: protect

File: dds/FACE/FaceTSS.cpp [hunk 7]
```diff
@@ -223,15 +237,22 @@ void Destroy_Connection(CONNECTION_ID_TYPE connection_id,
   OPENDDS_MAP(CONNECTION_ID_TYPE, Entities::FaceReceiver)& readers = entities.receivers_;
```
Comment (iamtheschmitzer) [inline dds/FACE/FaceTSS.cpp]: Use a typedref

File: dds/FACE/FaceTSS.cpp [hunk 8]
```diff
@@ -285,6 +308,214 @@ void Receive_Message(
...
+    Entities::ConnIdToSenderMap& writers = Entities::instance()->senders_;
+    Entities::ConnIdToSenderMap::iterator wtrIter = writers.begin();
+
+    while (wtrIter != writers.end()) {
+      temp_pub = wtrIter->second.dw->get_publisher();
+      temp_dp = temp_pub->get_participant();
+      DDS::PublisherQos temp_qos;
+      temp_pub->get_qos(temp_qos);
+      if (dp->get_domain_id() == temp_dp->get_domain_id() &&
+          temp_qos == qos) {
```
Comment (mitza-oci) [inline dds/FACE/FaceTSS.cpp]: using OpenDDS::DCPS::operator==;

**Pull request merged**